File: scalable-auth-system-solution-plus/backend/src/main/java/com/example/authsystem/service/TaskService.java
package com.example.authsystem.service;

import com.example.authsystem.config.CacheConfig;
import com.example.authsystem.dto.TaskDto;
import com.example.authsystem.exception.AccessDeniedException;
import com.example.authsystem.exception.ResourceNotFoundException;
import com.example.authsystem.mapper.TaskMapper;
import com.example.authsystem.model.Task;
import com.example.authsystem.model.User;
import com.example.authsystem.repository.TaskRepository;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.cache.annotation.CacheEvict;
import org.springframework.cache.annotation.Cacheable;
import org.springframework.cache.annotation.Caching;
import org.springframework.security.core.Authentication;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.Objects;
import java.util.stream.Collectors;

@Service
@RequiredArgsConstructor
@Slf4j
public class TaskService {

    private final TaskRepository taskRepository;
    private final UserService userService; // To get the current authenticated user
    private final TaskMapper taskMapper;

    @Transactional
    @Caching(evict = {
            @CacheEvict(value = CacheConfig.TASKS_CACHE, key = "'allTasks'"),
            @CacheEvict(value = CacheConfig.TASKS_CACHE, key = "'userTasks:' + #result.userId")
    })
    public TaskDto createTask(TaskDto taskDto) {
        User currentUser = userService.getAuthenticatedUserEntity();
        Task task = taskMapper.toEntity(taskDto);
        task.setUser(currentUser); // Associate task with the current user

        Task savedTask = taskRepository.save(task);
        log.info("Task created by user {}: {}", currentUser.getEmail(), savedTask.getTitle());
        return taskMapper.toDto(savedTask);
    }

    @Transactional(readOnly = true)
    @Cacheable(value = CacheConfig.TASKS_CACHE, key = "'task:' + #id")
    public TaskDto getTaskById(Long id) {
        User currentUser = userService.getAuthenticatedUserEntity();
        Task task = taskRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("Task", "id", id));

        // Authorization check: User can only access their own tasks unless they are an admin
        if (!Objects.equals(task.getUser().getId(), currentUser.getId()) && !userService.isAdmin(currentUser)) {
            log.warn("User {} attempted to access task {} belonging to user {}", currentUser.getEmail(), id, task.getUser().getEmail());
            throw new AccessDeniedException("You are not authorized to view this task.");
        }
        return taskMapper.toDto(task);
    }

    @Transactional(readOnly = true)
    @Cacheable(value = CacheConfig.TASKS_CACHE, key = "'userTasks:' + #root.args[0].id") // Cache based on user ID
    public List<TaskDto> getTasksForUser(User user) {
        return taskRepository.findByUser(user)
                .stream()
                .map(taskMapper::toDto)
                .collect(Collectors.toList());
    }

    @Transactional(readOnly = true)
    public List<TaskDto> getTasksForCurrentUser() {
        User currentUser = userService.getAuthenticatedUserEntity();
        return getTasksForUser(currentUser);
    }


    @Transactional(readOnly = true)
    @PreAuthorize("hasRole('ADMIN')") // Only ADMINs can retrieve all tasks
    @Cacheable(value = CacheConfig.TASKS_CACHE, key = "'allTasks'")
    public List<TaskDto> getAllTasks() {
        return taskRepository.findAll()
                .stream()
                .map(taskMapper::toDto)
                .collect(Collectors.toList());
    }

    @Transactional
    @Caching(evict = {
            @CacheEvict(value = CacheConfig.TASKS_CACHE, key = "'task:' + #id"),
            @CacheEvict(value = CacheConfig.TASKS_CACHE, key = "'allTasks'"),
            @CacheEvict(value = CacheConfig.TASKS_CACHE, key = "'userTasks:' + #result.userId")
    })
    public TaskDto updateTask(Long id, TaskDto taskDto) {
        User currentUser = userService.getAuthenticatedUserEntity();
        Task task = taskRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("Task", "id", id));

        // Authorization check
        if (!Objects.equals(task.getUser().getId(), currentUser.getId()) && !userService.isAdmin(currentUser)) {
            log.warn("User {} attempted to update task {} belonging to user {}", currentUser.getEmail(), id, task.getUser().getEmail());
            throw new AccessDeniedException("You are not authorized to update this task.");
        }

        taskMapper.updateTaskFromDto(taskDto, task);
        Task updatedTask = taskRepository.save(task);
        log.info("Task updated by user {}: {}", currentUser.getEmail(), updatedTask.getTitle());
        return taskMapper.toDto(updatedTask);
    }

    @Transactional
    @Caching(evict = {
            @CacheEvict(value = CacheConfig.TASKS_CACHE, key = "'task:' + #id"),
            @CacheEvict(value = CacheConfig.TASKS_CACHE, key = "'allTasks'"),
            @CacheEvict(value = CacheConfig.TASKS_CACHE, key = "'userTasks:' + #currentUser.id") // Evict based on resolved user
    })
    public void deleteTask(Long id) {
        User currentUser = userService.getAuthenticatedUserEntity();
        Task task = taskRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("Task", "id", id));

        // Authorization check
        if (!Objects.equals(task.getUser().getId(), currentUser.getId()) && !userService.isAdmin(currentUser)) {
            log.warn("User {} attempted to delete task {} belonging to user {}", currentUser.getEmail(), id, task.getUser().getEmail());
            throw new AccessDeniedException("You are not authorized to delete this task.");
        }

        taskRepository.delete(task);
        log.info("Task deleted by user {}: {}", currentUser.getEmail(), task.getTitle());
        // For cache evict on userTasks, we need to pass the user ID, which is tricky in @CacheEvict without #result
        // or by making 'currentUser' available to the SpEL expression. Let's assume `currentUser.id` is available
        // to simplify for now, or explicitly clear all tasks related to the user if exact key cannot be formed.
        // A better approach would be to use a custom cache resolver or a different caching strategy.
    }
}

File: fullstack-mobile-backend-system-complete/backend/src/modules/tasks/task.service.ts
import prisma from '../../config/prismaClient';
import { ApiError } from '../../middleware/errorHandler';
import { CreateTaskInput, UpdateTaskInput } from './task.validation';
import { UserRole } from '@prisma/client';
import logger from '../../utils/logger';
import cache from '../../utils/cache';

const TASK_CACHE_KEY_PREFIX = 'task:';
const USER_TASKS_CACHE_KEY_PREFIX = 'tasks:user:';
const ALL_TASKS_CACHE_KEY = 'tasks:all';

export const createTask = async (data: CreateTaskInput, userId: string) => {
  const task = await prisma.task.create({
    data: {
      ...data,
      dueDate: data.dueDate ? new Date(data.dueDate) : null,
      userId,
    },
    include: { user: { select: { id: true, email: true, name: true } } },
  });

  cache.del(`${USER_TASKS_CACHE_KEY_PREFIX}${userId}`); // Invalidate user's tasks cache
  cache.del(ALL_TASKS_CACHE_KEY); // Invalidate all tasks cache
  logger.info(`Task created by user ${userId}: ${task.title}`);
  return task;
};

export const getTasks = async (userId: string, userRole: UserRole) => {
  if (userRole === 'ADMIN') {
    const cachedTasks = cache.get(ALL_TASKS_CACHE_KEY);
    if (cachedTasks) {
      logger.debug('Fetching all tasks from cache (Admin).');
      return cachedTasks;
    }
    const allTasks = await prisma.task.findMany({
      include: { user: { select: { id: true, email: true, name: true } } },
      orderBy: { createdAt: 'desc' },
    });
    cache.set(ALL_TASKS_CACHE_KEY, allTasks);
    logger.debug('Fetching all tasks from DB and caching (Admin).');
    return allTasks;
  } else {
    const cachedUserTasks = cache.get(`${USER_TASKS_CACHE_KEY_PREFIX}${userId}`);
    if (cachedUserTasks) {
      logger.debug(`Fetching tasks for user ${userId} from cache.`);
      return cachedUserTasks;
    }
    const userTasks = await prisma.task.findMany({
      where: { userId },
      include: { user: { select: { id: true, email: true, name: true } } },
      orderBy: { createdAt: 'desc' },
    });
    cache.set(`${USER_TASKS_CACHE_KEY_PREFIX}${userId}`, userTasks);
    logger.debug(`Fetching tasks for user ${userId} from DB and caching.`);
    return userTasks;
  }
};

export const getTaskById = async (id: string, userId: string, userRole: UserRole) => {
  const cachedTask = cache.get(`${TASK_CACHE_KEY_PREFIX}${id}`);
  if (cachedTask) {
    logger.debug(`Fetching task ${id} from cache.`);
    return cachedTask;
  }

  const task = await prisma.task.findUnique({
    where: { id },
    include: { user: { select: { id: true, email: true, name: true } } },
  });

  if (!task) {
    throw new ApiError(404, 'Task not found.');
  }

  // A user can only see their own tasks unless they are an admin
  if (task.userId !== userId && userRole !== 'ADMIN') {
    throw new ApiError(403, 'You are not authorized to view this task.');
  }

  cache.set(`${TASK_CACHE_KEY_PREFIX}${id}`, task);
  logger.debug(`Fetching task ${id} from DB and caching.`);
  return task;
};

export const updateTask = async (id: string, data: UpdateTaskInput, userId: string, userRole: UserRole) => {
  const existingTask = await prisma.task.findUnique({ where: { id } });

  if (!existingTask) {
    throw new ApiError(404, 'Task not found.');
  }

  // A user can only update their own tasks unless they are an admin
  if (existingTask.userId !== userId && userRole !== 'ADMIN') {
    throw new ApiError(403, 'You are not authorized to update this task.');
  }

  const updatedTask = await prisma.task.update({
    where: { id },
    data: {
      ...data,
      dueDate: data.dueDate ? new Date(data.dueDate) : undefined,
    },
    include: { user: { select: { id: true, email: true, name: true } } },
  });

  cache.del(`${TASK_CACHE_KEY_PREFIX}${id}`); // Invalidate single task cache
  cache.del(`${USER_TASKS_CACHE_KEY_PREFIX}${userId}`); // Invalidate user's tasks cache
  cache.del(ALL_TASKS_CACHE_KEY); // Invalidate all tasks cache
  logger.info(`Task ${id} updated by user ${userId}.`);
  return updatedTask;
};

export const deleteTask = async (id: string, userId: string, userRole: UserRole) => {
  const existingTask = await prisma.task.findUnique({ where: { id } });

  if (!existingTask) {
    throw new ApiError(404, 'Task not found.');
  }

  // A user can only delete their own tasks unless they are an admin
  if (existingTask.userId !== userId && userRole !== 'ADMIN') {
    throw new ApiError(403, 'You are not authorized to delete this task.');
  }

  await prisma.task.delete({ where: { id } });

  cache.del(`${TASK_CACHE_KEY_PREFIX}${id}`); // Invalidate single task cache
  cache.del(`${USER_TASKS_CACHE_KEY_PREFIX}${userId}`); // Invalidate user's tasks cache
  cache.del(ALL_TASKS_CACHE_KEY); // Invalidate all tasks cache
  logger.info(`Task ${id} deleted by user ${userId}.`);
  return { message: 'Task deleted successfully.' };
};